pkg/virt-api/rest/streamer.go

```diff
@@ -0,0 +1,162 @@
+type streamFuncResult error
+
+type Streamer struct {
+	fetchVMI        vmiFetcher
+	validateVMI     validator
+	dial            dialer
+	keepAliveClient func(ctx context.Context, conn *websocket.Conn, cancel func())
+
+	streamToClient streamFunc
+	streamToServer streamFunc
+}
+
+func NewRawStreamer(fetch vmiFetcher, validate validator, dial dialer) *Streamer {
+	return &Streamer{
+		fetchVMI:    fetch,
+		validateVMI: validate,
+		dial:        dial,
+		streamToServer: func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			_, err := io.Copy(serverConn, clientConn.UnderlyingConn())
+			result <- err
+		},
+		streamToClient: func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			_, err := io.Copy(clientConn.UnderlyingConn(), serverConn)
+			result <- err
+		},
+	}
+}
+
+func NewWebsocketStreamer(fetch vmiFetcher, validate validator, dial dialer) *Streamer {
+	return &Streamer{
+		fetchVMI:    fetch,
+		validateVMI: validate,
+		dial:        dial,
+		streamToServer: func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			_, err := kubecli.CopyFrom(serverConn, clientConn)
+			result <- err
+		},
+		streamToClient: func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			_, err := kubecli.CopyTo(clientConn, serverConn)
+			result <- err
+		},
+	}
+}
+
+func (s *Streamer) Handle(request *restful.Request, response *restful.Response) error {
+	namespace := request.PathParameter(NamespaceParamName)
+	name := request.PathParameter(NameParamName)
+
+	vmi, statusErr := s.fetchAndValidateVMI(namespace, name)
+	if statusErr != nil {
+		writeError(statusErr, response)
+		return statusErr
+	}
+
+	serverConn, statusErr := s.dial(vmi)
+	if statusErr != nil {
+		writeError(statusErr, response)
+		return statusErr
+	}
+	clientConn, err := clientConnectionUpgrade(request, response)
+	if err != nil {
+		writeError(errors.NewBadRequest(err.Error()), response)
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(request.Request.Context())
+	defer cancel()
+	go s.cleanupOnClosedContext(ctx, clientConn, serverConn)
+
+	if s.keepAliveClient != nil {
+		go s.keepAliveClient(context.Background(), clientConn, cancel)
+	}
+
+	results := make(chan streamFuncResult, 2)
+	defer close(results)
+
+	go s.streamToClient(clientConn, serverConn, results)
+	go s.streamToServer(clientConn, serverConn, results)
+
+	result1 := <-results
+	// start canceling on the first result to force all goroutines to terminate
+	cancel()
+	result2 := <-results
+
+	if result1 != nil {
+		return result1
+	}
+	return result2
+}
+
+func (s *Streamer) fetchAndValidateVMI(namespace, name string) (*v1.VirtualMachineInstance, *errors.StatusError) {
+	vmi, err := s.fetchVMI(namespace, name)
+	if err != nil {
+		return nil, err
+	}
+	if err := s.validateVMI(vmi); err != nil {
+		return nil, err
+	}
+	return vmi, nil
+}
+
+const streamTimeout = 10 * time.Second
```

Comment:
Not sure if this is to short or to long. This means that if the client is unreachable for 10s the tunnel will get closed.

Reply:
Good question. Since we can't make the timeout right now easily configurable, maybe we should be more graceful? I guess for the handshake timeout 10 seconds is good. For the ping/pong timeout, maybe one minute would be better?